[Stevan Gerard Gunawan] iP

gerardstevan opened the pull request:
# Doke

Doke is a simple java program which would keep track of Tasks. It would utilize user inputs to produce outputs and tasks and any tasks would be stored locally in your device after every addition or changes.

It would be able to handle Task such as:

- Events
- To-Dos
- Deadlines

Features: 

- [X] Managing Task
- [x] Keep deadline and Event date
- [ ]  Further expansion of task types and features (coming soon)

All you need to do is,

1. Pull it from [here ](https://github.com/gerardstevan/ip.git)
2. Open it in your IDE or terminal.
3. compile and run the Doke.java file.
4. let it manage your **tasks** for you :upside_down_face:

If you are doing it in terminal, you can go to the directory which contains the code and enter `javac Doke.java` before entering `java Doke` if you have java installed in your device and the program should run. 
The following is the body of the main method in Doke.java.

``` java
        Ui ui = new Ui();
        Scanner sc = new Scanner(System.in);
        Storage storage = new Storage();
        Parser parser = new Parser();
        TaskList taskList;

        ui.printOut("full path: " + Storage.DOKE_FILE.getAbsolutePath());

        ui.printOut("hello");

        taskList = new TaskList(ui, storage);
        ui.printOut("The following is your stored task: \n");
        taskList.listOut(ui);

        String str = sc.nextLine();
        while (true) {
            if (!parser.processString(taskList, storage, str, ui)) {
                break;
            }
            str = sc.nextLine();
        }

        ui.printOut("bye");
```

If you are interested in contributing to the project do keep in mind this quote:

> “Always code as if the guy who ends up maintaining your code will be a violent psychopath who knows where you live”
> -John Woods

Commit:
Add Gradle support

gerardstevan committed:
Level-1

gerardstevan committed:
Level-2

gerardstevan committed:
Level-3

gerardstevan committed:
Level-4

gerardstevan committed:
A-TextUiTesting

gerardstevan committed:
Level-5

gerardstevan committed:
Level-6

gerardstevan committed:
improved Level-6

gerardstevan committed:
Improve code structure

gerardstevan committed:
fixed a minor bug

gerardstevan committed:
Level-7

gerardstevan committed:
implement Level-8

gerardstevan committed:
Merge branch 'branch-Level-7'

gerardstevan committed:
Merge branch 'branch-Level-8'

gerardstevan committed:
minor bug fix

gerardstevan committed:
Implement A-MoreOOP.

gerardstevan committed:
organize classes into doke package

gerardstevan committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

gerardstevan committed:
set up gradle

gerardstevan committed:
Add J-unit tests

gerardstevan committed:
change jar file name to doke.jar

gerardstevan committed:
Add javadocs

gerardstevan committed:
Implement cs2103t  coding standard

gerardstevan committed:
Implement find feature

The find feature will try to find the keyword in the Task's desc and
time if it is an Event or Deadline task.

gerardstevan committed:
Merge branch 'branch-A-JavaDoc'

gerardstevan committed:
Merge branch 'branch-A-CodingStandard'

gerardstevan committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/doke/Parser.java
#	src/main/java/doke/TaskList.java

gerardstevan committed:
Improve stringSearch to not differentiate between upper and lower case

gerardstevan committed:
Added simple GUI

gerardstevan committed:
Merge branch 'branch-Level-10'

rui-han-crh commented on src/main/java/doke/DialogBox.java:
```diff
@@ -0,0 +1,47 @@
+        this.getChildren().setAll(tmp);
+    }
+
+    public static DialogBox getUserDialog(Label l, ImageView iv) {
```
Perhaps consider a more explicity naming convention? l and iv may not convey the variable well unless the reader refers to the type in the formal parameter.

rui-han-crh replied:
Oh I see, I think this is the default names from the guide 😅

rui-han-crh commented on src/main/java/doke/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     * @param ui
+     * @param storage
+     */
+    public TaskList(Ui ui, Storage storage) {
```
This constructor is slightly long and the control flow may be hard to follow due to some inverted statements.
Perhaps consider shifting the Scanner responsibilities to a new method/class?

rui-han-crh commented on src/main/java/doke/Events.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Represent the Events task. Extends the Task class.
+ */
+public class Events extends Task{
```
Perhaps the name of this class should be `Event` to signify a singular object
```suggestion
public class Event extends Task{
```

rui-han-crh commented on src/main/java/doke/Events.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + at.format(DateTimeFormatter.ofPattern("dd-MMM-yyyy")) + ")";
```
Perhaps consider caching the `DateTimeFormatter::ofPattern` output once during construction to reduce load of toString calls?

rui-han-crh commented on src/main/java/doke/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+        String message = "_________________________ \n";
+        int i = 0;
+        message = "_________________________ \n";
+        while (i < len) {
```
It seems like i is incremented at the end of the loop no matter what, perhaps the use of the enhanced for loop will make the purpose of this loop more explicit?

rui-han-crh commented on src/test/java/doke/TodoTest.java:
```diff
@@ -0,0 +1,20 @@
+public class TodoTest {
+
+    @Test
+    public void getType_emptyString_stringTReturned() {
```
I like the naming convention of these test cases - they are very clear in their domain and expected outcomes

rui-han-crh commented on src/main/java/doke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+     * @param ui to print out messages
+     * @return boolean to signify whether the doke program should end or not.
+     */
+    public boolean processString(TaskList taskList, Storage storage, String string, Ui ui) {
```
This function seems to be more than 100 lines long and explicitly declares the implementations of all parsed instructions. Perhaps this does not conform well to the SLAP principle and would benefit from abstracting and splitting each process in another private method and then calling these methods to improve readability and maintainability?

rui-han-crh commented on src/main/java/doke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+     * @return a string representation of the task which can be written to the storage file.
+     */
+    public static String createWordForFile(Task task) {
+        String time = task instanceof ToDo ? "" : " | " + task.getTime();
```
I like the use of the ternary operator here

rui-han-crh commented on src/main/java/doke/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    /**
+     * Returns a copy of the ArrayList.
+     *
+     * @return a copy of the ArrayList.
+     */
+    public ArrayList<Task> getList() {
+        ArrayList<Task> temp = this.taskList;
+        return temp;
+    }
```
There seems to be an error in this function, it mentions that it would like to return a copy of the ArrayList, but the current implementation will result in aliasing the list. Perhaps consider cloning and casting the result?

```suggestion
     * Returns a copy of the ArrayList.
     *
     * @return a copy of the ArrayList.
     */
    public ArrayList<Task> getList() {
        ArrayList<Task> temp = (ArrayList<Task>)this.taskList.clone();
        return temp;
    }
```

Although, I'm unsure this if this a documentation error. If so, maybe it would be a better choice to describe the function as `Retrieves the current TaskList`, as the word `copy` may indicate the explicit acting of copying over the objects to a new collection.

rui-han-crh commented on src/main/java/doke/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+    public TaskList(Ui ui, Storage storage) {
+
+        try {
+            Scanner s = new Scanner(Storage.DOKE_FILE);
```
I like that you opened a scanner and consumed the `DOKE.FILE`, this is a neat deserialisation method!

rui-han-crh reviewed (commented):
I like that there we clearly defined test cases with explicit output results.

A greater use of abstraction could greatly enhance the potential for readability! 👍

sunyuting2000 commented on src/main/java/doke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+ * a class to represents the storage.
+ */
+public class Storage {
+    protected static final File DOKE_FILE = new File(Doke.DOKE_FILE_PATH);
```
All capitalized name suggests that it is a constant instead of a variable. Good.

sunyuting2000 commented on src/main/java/doke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param arrayList containing the task to be written to the file.
+     */
+    public void updateFile(ArrayList<Task> arrayList, Ui ui) {
```
The methods are all verbs and are all in camelCase. Good.

sunyuting2000 commented on src/main/java/doke/ToDo.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Represent the ToDo task. Extends the Task class.
+ */
+public class ToDo extends Task{
```
There should be a space between the Task and {, like so

```suggestion
public class ToDo extends Task {
```
Other than that nice job!

Effixion reviewed (commented):
Overall there were not many coding standards that were breached. Good job!

gerardstevan committed:
Added FXML

gerardstevan committed:
Refactor code to improve code quality

Extracted method from some arrowhead method in taskList and and parser,
especially with the help of the methods in the new Command class

gerardstevan committed:
Refactor code to follow checkstyle and remove arrowhead method.

Fixed the code style problems suggested by checkstyle and refactored the
method for searchString to remove the existing arrowhead.

gerardstevan committed:
Add assertion to a few methods which uses assumption

Assertion are added to methods in DokeException and TaskList because of
some assumptions made in the method, which if violated may throw an
exception.

gerardstevan committed:
Merge pull request #1 from gerardstevan/branch-A-CodeQuality

Improve code quality

gerardstevan committed:
Revert "Branch-A-CodeQuality"

gerardstevan committed:
Merge pull request #2 from gerardstevan/revert-1-branch-A-CodeQuality

Revert "Branch-A-CodeQuality"

gerardstevan committed:
Merge pull request #3 from gerardstevan/branch-A-Assertions

Add assertion to a few methods which uses assumption

gerardstevan committed:
Made a minor change to improve readability

gerardstevan committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/doke/Command.java

gerardstevan committed:
Remove a debug print statement

Added assert statements which was lost in the previous merge

gerardstevan committed:
Merge pull request #5 from gerardstevan/branch-A-CodeQuality

Branch-A-CodeQuality

gerardstevan committed:
Implement sort feature for tasks

Added a feature which will sort (C-Sort) the taskList based on date,
and todo tasks will be at the end of the list. This happens when sort
is entered on the Command Line.

gerardstevan committed:
Made the GUI functional

Connect the GUI with the program logic so it behaves as wanted by
refactoring some part of the code. Fixed some checkstyle issue.

gerardstevan committed:
Handle error for invalid integer for deleteCommand

Handle error for invalid integer for deleteCommand.
Fixed a minor GUI issue.

gerardstevan committed:
Update the User Guide.

Updated the user guide in docs/README.md. Added a picture of the app
in docs/Ui.png. Fixed a minor GUI bug.

gerardstevan committed:
Fix a minor GUI bug

Made the GUI functional for doke.jar. Improve the GUI response

gerardstevan committed:
Fix a minor GUI bug.

Fixed the bug where the GUI didn't exit when `bye` command is entered.
Prevent the GUI app from resizing.
